Apps/Src/ReadCarCAN.c

```diff
@@ -68,110 +75,61 @@ static void updateSaturation(int8_t chargeMessage){
-
-            restartingArray = true;
-
-            // Wait to make sure precharge is finished and then restart array
-            OSTmrStart(&prechargeDlyTimer, &err);
-            assertOSError(OS_READ_CAN_LOC, err);
-            
-        }
-    
-
-    OSMutexPost(&arrayRestartMutex, OS_OPT_NONE, &err);
-    assertOSError(OS_READ_CAN_LOC,err);
+/**
+ * @brief Nested function as the same function needs to be executed however the timer requires different parameters
+ * @param p_tmr pointer to the timer that calls this function, passed by timer
+ * @param p_arg pointer to the argument passed by timer
+*/
+static void assertMissedBPSMsg(void *p_tmr, void *p_arg){
+    assertReadCarCANError(READCARCAN_ERR_MISSED_MSG);
 }
 
 /**
  * @brief Callback function for the precharge delay timer. Waits for precharge and then restarts the array.
  * @param p_tmr pointer to the timer that calls this function, passed by timer
  * @param p_arg pointer to the argument passed by timer
- * 
 */
 static void arrayRestart(void *p_tmr, void *p_arg){
-    Minion_Error_t Merr;
     if(chargeEnable){    // If regen has been disabled during precharge, we don't want to turn on the main contactor immediately after
-        Contactors_Set(ARRAY_CONTACTOR, (Minion_Read_Pin(IGN_1, &Merr)), false); //turn on array contactor if the ign switch lets us
-        UpdateDisplay_SetArray(true);
-    }
-    // done restarting the array
-    restartingArray = false;
+        Contactors_Set(ARRAY_CONTACTOR, arrayIgnitionStatus, false); // Turn on array contactor if the ign switch lets us
 
-};
-
-/**
- * @brief Disables charging when canWatchTimer hits 0 (when charge enable messages are missed)
- * @param p_tmr pointer to the timer that calls this function, passed by timer
- * @param p_arg pointer to the argument passed by timer
-*/
-void canWatchTimerCallback (void *p_tmr, void *p_arg){
-    chargingDisable();
-}
+        // Update display based on if array contactor is on/off
+        if(Contactors_Get(ARRAY_CONTACTOR) == ON){
+            UpdateDisplay_SetArray(true);
+        }else{
+            UpdateDisplay_SetArray(false);
+        }
 
+        #ifdef __TEST_READCARCAN
+        printf("\r\nArray Contactor is turned on");
+        #endif
 
+    }
+};
 
-void Task_ReadCarCAN(void *p_arg)
-{
+void Task_ReadCarCAN(void *p_arg){
     OS_ERR err;
 
-    //data struct for CAN message
+    // data struct for CAN message
     CANDATA_t dataBuf;
 
-    OSMutexCreate(&arrayRestartMutex, "array restart mutex", &err);
-    assertOSError(OS_READ_CAN_LOC,err);
-
-
     // Create the CAN Watchdog (periodic) timer, which disconnects the array and disables regenerative braking
     // if we do not get a CAN message with the ID Charge_Enable within the desired interval.
     OSTmrCreate(
         &canWatchTimer,
         "CAN Watch Timer",
-        0,
-        CAN_WATCH_TMR_DLY_TMR_TS,
+        CAN_WATCH_TMR_DLY_TMR_TS, // Initial delay equal to the period since 0 doesn't seem to work
```

Comment:
Zero should work. Are you sure about this?

Reply:
@IshDeshpa @Cam0Cow and I tested it with 0 on hardware yesterday but it didn't seem to work. It only worked with that specific delay.